### ISSUE TITLE
When extension is omitted, IO.popen does not find the executable in JRuby on Windows

### ISSUE THREAD

**Issue by jcouball** [jruby/jruby#6516]
**Environment Information**
  
JRuby version and operating system:
* jruby 9.2.14.0 (2.5.7) 2020-12-08 ebe64bafb9 OpenJDK 64-Bit Server VM 25.275-b01 on 1.8.0_275-b01 +jit [mswin32-x86_64]
* jruby 9.3.0.0-SNAPSHOT (2.6.5) 2020-12-28 08642361e6 OpenJDK 64-Bit Server VM 25.275-b01 on 1.8.0_275-b01 +jit [mswin32-x86_64]

Other relevant info:
* I think this is another instance of #5018

**Expected Behavior**

GIVEN the following Ruby code running in JRuby on Windows

AND the following statement succeeds:
```Ruby
output = IO.popen(["gem.bat", "build", "-V", "jruby_test.gemspec"], &:read)
```

AND the PATHEXT environment variable is set to `.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL`

Then following statement (which omits the ".bat" extension) should also succeed:
```Ruby
output = IO.popen(["gem", "build", "-V", "jruby_test.gemspec"], &:read)
```

I noticed this in my CI job for my gem.  I do not have a windows box to test this on, but was able to duplicate this using GitHub Actions using a Windows VM:

https://github.com/jcouball/jruby_test/blob/main/Rakefile#L7

In JRuby on Windows, the `popen_test_1` rake task succeeds and the `popen_test_2` rake task fails.

**Actual Behavior**

The second command fails with the error:

```
IOError: Cannot run program "gem" (in directory "D:\a\jruby_test\jruby_test"): CreateProcess error=2, The system cannot find the file specified
```

**Comment by headius** [comment]
Aha, so during the executable search on Windows we should be trying PATHEXT extensions (in sequence? all of them?) to see if they exist (assuming the filename as given does not resolve to an executable... or perhaps always if it does not end with a PATHEXT value?)

Unfortunately the Windows process logic currently still uses the JDK and does not utilize the large piece of ported process-launching code from CRuby. We will want to change that or otherwise have to make this fix in the ShellLauncher JDK process wrapper garbage.

**Comment by jcouball** [comment]
Thank you for the response.

I am not certain what the right answers are to your questions. You probably want the same behavior as MRI Ruby, so use that as your guide to answer these questions.

In my example project, I get the expected behavior in MRI Ruby on Windows:
https://github.com/jcouball/jruby_test

**Comment by headius** [comment]
@jcouball No worries, I think we can take it from here.

Perhaps you could check if the Ruby specs (https://github.com/ruby/spec) has any cases to test this in, for example, core/process/spawn_spec.rb?

**Comment by jcouball** [comment]
@headius I studied the ruby/spec and ruby/ruby repositories and couldn't find test cases covering this.  I tried following the Ruby implementation code code to see where this is implemented but there are so many #ifdefs it was super hard to follow what was going on. I suspect that this isn't implemented in the Ruby code at all, but a result of some system call.

**Comment by headius** [comment]
@jcouball You findings about the implementation match mine... I could not find anywhere that they explicitly try to assemble the filename plus extensions. You may be right about this being a standard part of the process launching native calls they use on Win32.

The next thing to look into would be to see how CRuby determines that there will be an executable command **before** it does the process call. A couple answers are possible:

* They may do no executable check at all on Windows and just allow the process launch to fail with an appropriate error.
* They may be calling a specific Win32 function to pre-check for an executable.

It may be that we should just remove our pre-check of the executable and allow it to fail, since we do not have a lot of native control over process launching on Windows (yet).

**Comment by jcouball** [comment]
I think I found where the executable filename is assembled: [dln_find_exe_r in dln_find.c](https://github.com/ruby/ruby/blob/590dc06e3840cc7b00d80ccaac9fbf42573428f8/dln_find.c#L62).

I think this is only called when Ruby determines that it should run the command directly instead of passing the command to the shell for execution.   The `rb_execarg_new` function has logic to determine if the shell will be called to execute the command or not.  This function is called from all/most of the Process functions that create a subprocess like spawn, exec, and system.

**Comment by headius** [comment]
@jcouball Aha, well searched! So they do have logic to apply the `EXECUTABLE_EXTS` to the filename, in turn, until one of them exists. This will help us do the same in JRuby, thank you!

**Comment by jcouball** [comment]
Ok, I'll leave the rest to you and your team!  Thanks for looking into this.

For additional context, I originally ran into this problem when trying to run the Rake `build` task from `bundler` on Windows.  As a workaround, I redefined the `build` and `install` tasks to avoid using `popen`:

https://github.com/ruby-git/ruby-git/blob/d1b17118ff69bf6913ddb27a7f2ffbb77d639da0/Rakefile#L45-L69

**Comment by headius** [comment]
@jcouball If I can get up the motivation I will attempt to get process launching on Windows to use the same logic as on UNIX. If that can be done then this should be an easy port. If not, less easy but the logic will be the same. Whatever we manage to do will be in 9.3, due out in the next month or so.

**Comment by headius** [comment]
It appears to work properly for `system` so we may be able to reuse the same logic:

```
P:\jruby>bin\jruby -e "system 'bin/gem', '-v'"
3.0.6

P:\jruby>bin\jruby -e "IO.popen ['bin/gem', '-v']; sleep 5"
IOError: Cannot run program "bin\gem" (in directory "P:\jruby"): CreateProcess error=2, The system cannot find the file specified
   popen at org/jruby/RubyIO.java:4092
  <main> at -e:1
```

**Comment by headius** [comment]
Fixes being verified now in #6532.

**Comment by headius** [comment]
Verified by @enebo. Fixed!
